### ISSUE TITLE
Refactoring ng-bootstrap to elimiminate dependence on (ng4 deprecated) `<template>`

### ISSUE THREAD

**Issue by somombo** [ng-bootstrap/ng-bootstrap#1344]
The problematic `<template>`  tag has been deprecated in ng4. See https://github.com/ng-bootstrap/ng-bootstrap/issues/1337

So It seems a good a time as any, to discuss what seems to me to be over usage of `<template>` throughout the ng-bootstrap project. 

> Note: I have put togther a PR addressing this issue for carousel https://github.com/ng-bootstrap/ng-bootstrap/pull/1343

Is doing something like this something we might consider for the rest of the project?

I believe we should for the following reasons:

I think it's not a very good sign, especially when even `ng-bootstrap` users are forced to use it, as is the case with `ngbSlide`. I think its possible to refactor this project in a manner that's more [idomatic to angular](https://angular.io/docs/ts/latest/guide/structural-directives.html#!#prefer-asterisk). 

Its often hard to follow `template` code and I'm not sure why it's being used so often where there are simpler and more conventional ways of expressing things.

 For example in places where there is:

```html
<template ngFor [ngForOf]="parts" let-part let-isOdd="odd">
```
 one could simply use,
```html
<ng-container *ngFor="let part of parts; let isOdd=odd">
```

And instead of `<template [ngIf]="seconds">`, the directive `<ng-container *ngIf="seconds">` should be used instead. 

Also, I think in alot of cases its also possible to do away with `<template [ngTemplateOutlet]="...">` in favor of somehow making use of `<ng-content>` or `<ng-container>` [(which is already officially encouraged in angular docs)](https://angular.io/docs/ts/latest/guide/structural-directives.html#!#ng-container).

I think this is an important step as each of the latter cases are already supported by both angular 2 and angular 4(+); whereas `<template>` is ng2 only and `<ng-template>` will be in ng4 only. So `template` (or even `ng-template`) should only be used as a last resort. And even if it is used for under-the-hood mechanics, I think front facing API shouldn't require one to use `<template>` as is the case in the carousel component.

What do you all think?

**Comment by pkozlowski-opensource** [comment]
@somombo we've seen [your comment](https://github.com/ng-bootstrap/ng-bootstrap/issues/1337#issuecomment-282555634) in #1337 - no need to open a separate issue.

tl;dr; is that `<template>` and `<ng-content>` got _different_ semantics so you can't simply swap one for another. Will respond with more details in #1337

Closing as a duplicate of #1337

**Comment by somombo** [comment]
@pkozlowski-opensource I removed my comment from https://github.com/ng-bootstrap/ng-bootstrap/issues/1337 so this is nolonger duplicate

**Comment by somombo** [comment]
@pkozlowski-opensource I think its more appropriate for this to be an issue on its own because its not simply about the fact that `<template>` has been deprecated (as is the case in #1337).  And in fact I made this an issue on its own because I am aware that what I'm addressing will not be a simple swap as is the case in issue https://github.com/ng-bootstrap/ng-bootstrap/issues/1337

I think if we go along the lines of just swapping `ng-template` for `template` then we will miss out on an oppotunity to do some needed cleanup and will simply be trading one problem for another

**Comment by pkozlowski-opensource** [comment]
@somombo let me respectfully disagree:
* #1337 is IMO right place to tackle things as we need to find replacement for `<template>`
* you seem to be implying that there is a problem with using `<template>` / `<ng-template>` and I wholeheartedly disagree - usage of  `<template>` / `<ng-template>` was / is conscious design decision that was discussed at length with Misko and material teams. 

Once again, there is semantic difference between using  `<template>` / `<ng-template>`  and `<ng-content>` / components: when `<template>` is used directives and change detection only kicks in when corresponding view is inserted into the DOM. In case of `<ng-content>` all the views and directives are instantiated and change-detected even if not used. This is _not_ something we want to have for components like accordion / tabset (I guess it matters less for a carousel where usually there is a limited number of "lightweight" slides).

I think that we could have more productive discussion if you stop assuming that usage of  `<template>` / `<ng-template>` is problematic - there are just different trade-offs and we've decided to use `<template>` / `<ng-template>` to get best performance by default.

**Comment by somombo** [comment]
I think it's better to leverage use of `ChangeDeteaction.OnPush` (and e.g. observables)  to gain  finer grain control over changes. But if you say all this has been considered then I concede. Thanks for your timely response.

**Comment by somombo** [comment]
@pkozlowski-opensource If you still feel there is any merit to PR https://github.com/ng-bootstrap/ng-bootstrap/pull/1343 (besides what you've mentioned), then feel free to use it  otherwise you may discard it
